File: 0249-group-shifted-strings/0249-group-shifted-strings.cpp
class Solution {
public:
    string findKey(string& s){
        string ret = ""; 
        ret += '0'; 
        char curr = s[0]; 
        for(int i = 1; i < s.length(); i++){
            int diff = s[i] - curr < 0 ? (s[i] - curr) + 26 : (s[i] - curr); 
            curr = s[i];  
            ret += (diff + '0'); 
        }
        return ret; 
    }
    vector<vector<string>> groupStrings(vector<string>& strings) {
        vector<vector<string>> answer; 
        unordered_map<string,vector<string>> hashMap; 
        for(string& s : strings){
            string key = findKey(s); 

            hashMap[key].push_back(s); 
        }
        
        for(auto& it : hashMap){
            answer.push_back(it.second); 
        }

        return answer; 
    }
};

File: 0249-group-shifted-strings/0249-group-shifted-strings.java
class Solution {
    public List<List<String>> groupStrings(String[] strings) {
        List<List<String>> answer = new ArrayList<>(); 
        Map<String,List<String>> hashMap = new HashMap<>();
        
        for(String s : strings){
            String key = convert(s); 
            hashMap.computeIfAbsent(key, k -> new ArrayList<>()).add(s); 
        }

        for(var e : hashMap.entrySet()){
            answer.add(e.getValue()); 
        }

        return answer; 
    }

    public String convert(String s){
        String key = "1"; 
        for(int i = 1; i < s.length(); i++){
            int diff = (s.charAt(i) - s.charAt(i-1) + 26) % 26;   // 이 줄만 변경
            key += String.valueOf(diff) + ","; 
        }
        //System.out.println(key); 
        return key; 
    }
}